Implement Error for DecodeError and BufReadDecoderError.

dstu opened the pull request:
This is a basic impl of Error (and Display). It should address #4.

Commit:
Implement Error for DecodeError and BufReadDecoderError.

wfraser commented on src/read.rs:
```diff
@@ -30,6 +33,24 @@ impl<'a> BufReadDecoderError<'a> {
+impl<'a> fmt::Display for BufReadDecoderError<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+      match *self {
+          BufReadDecoderError::InvalidByteSequence(_) => write!(f, "invalid byte sequence"),
```
consider also including what the bad bytes are:
```suggestion
          BufReadDecoderError::InvalidByteSequence(bytes) => write!(f, "invalid byte sequence {:x?}", bytes),
```

dstu replied:
Good call. I also added some padding and reformatted a tiny bit, to respect rustfmt's default line length preferences.

Commit:
InvalidByteSequence should share what that byte sequence is.

I also hit the affected hunk with 'cargo fmt' because the lines were getting a
little long.

Commit:
Provide more comprehensive Display impls for DecodeError, too.

SimonSapin commented on src/lib.rs:
```diff
@@ -28,6 +30,25 @@ pub enum DecodeError<'a> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            DecodeError::Invalid { valid_prefix, invalid_sequence, remaining_input, } => {
+                write!(f, "invalid byte sequence (preceding valid string of length {}: '{}'; \
```
I parse this as "invalid sequence that precedes a valid string, which is backward. Could you rephrase this? Also, maybe reduce repetition. Something like: `found invalid byte sequence {:02x?} after {} valid bytes, followed by {} more unprocessed bytes`

(Similarly below for incomplete)

dstu replied:
Done. I also hit it with `cargo fmt` again (which is the reason for the syntactic churn).

SimonSapin commented:
Hi! Thanks for the PR. If you’d like this change on crates.io, please also increment the version number in `Cargo.toml`.

Commit:
Try to make the Display impls for error variants easier to read.

Commit:
Minor version number bump for recent extensions to error types.

dstu commented:
Happy to contribute. Thanks for the review.

SimonSapin commented:
Thanks!
